subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_rai.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/socket.h>
+#include <zephyr/net/lwm2m.h>
+#include <modem/lte_lc.h>
+#include <nrf_modem_at.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(lwm2m_rai, CONFIG_LWM2M_CLIENT_UTILS_LOG_LEVEL);
+static void lte_rrc_event_handler(const struct lte_lc_evt *const evt);
+
+static bool rrc_connected;
+
+static int set_rel14feat(void)
+{
+	int ret;
+
+	ret = nrf_modem_at_printf("AT%%REL14FEAT=0,1,0,0,0");
+	if (ret) {
+		LOG_ERR("AT command failed, error code: %d", ret);
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static int read_rai(void)
+{
+	uint16_t rai;
+	int ret;
+
+	ret = nrf_modem_at_scanf("AT%RAI?", "%%RAI: %hu", &rai);
+	if (ret != 1) {
+		LOG_ERR("Failed to read RAI, error code: %d", ret);
+		return ret;
+	}
+	LOG_DBG("RAI: %d", rai);
+	return 0;
+}
+
+static int rai_req(bool enable)
+{
+	int ret;
+
+	ret = nrf_modem_at_printf("AT%%RAI=%d", enable);
+	if (ret) {
+		LOG_ERR("AT command failed, error code: %d", ret);
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static int lwm2m_lte_handler_register(void)
+{
+	lte_lc_register_handler(lte_rrc_event_handler);
+
+	return 0;
+}
+
+static void lte_rrc_event_handler(const struct lte_lc_evt *const evt)
+{
+	switch (evt->type) {
+	case LTE_LC_EVT_RRC_UPDATE:
+		if (evt->rrc_mode == LTE_LC_RRC_MODE_CONNECTED) {
+			rrc_connected = true;
+		} else {
+			rrc_connected = false;
+		}
+		LOG_DBG("LTE RRC mode: %s", evt->rrc_mode == LTE_LC_RRC_MODE_IDLE ? "Idle" :
+					    evt->rrc_mode == LTE_LC_RRC_MODE_CONNECTED ?
+										    "Connected" :
+										    "Unknown");
+		break;
+	default:
+		break;
+	}
+}
+
+int lwm2m_rai_no_data(void)
+{
+	int ret;
+	struct lwm2m_ctx *ctx;
+
+	if (rrc_connected) {
+		ctx = lwm2m_rd_client_ctx();
+
+		if (ctx->sock_fd >= 0) {
+			LOG_DBG("Set socket option SO_RAI_NO_DATA");
+			ret = setsockopt(ctx->sock_fd, SOL_SOCKET, SO_RAI_NO_DATA, NULL, 0);
+
+			if (ret < 0) {
+				ret = -errno;
+				LOG_ERR("Failed to set RAI socket option, error code: %d", ret);
+				return ret;
+			}
+		}
+	}
+
+	return 0;
+}
+
+int lwm2m_rai_last(void)
+{
+	int ret;
+	struct lwm2m_ctx *ctx;
+	u_int8_t data = 0;
+
+	if (rrc_connected) {
+		ctx = lwm2m_rd_client_ctx();
+
+		if (ctx->sock_fd >= 0) {
+			LOG_INF("Set socket option SO_RAI_LAST");
+			ret = setsockopt(ctx->sock_fd, SOL_SOCKET, SO_RAI_LAST, NULL, 0);
+
+			if (ret < 0) {
+				ret = -errno;
+				LOG_ERR("Failed to set RAI socket option, error code: %d", ret);
+				return ret;
+			}
+
+			ret = send(ctx->sock_fd, &data, sizeof(data), 0);
```

Comment:
Do we need to send a payload? Can we send just an empty UDP with payload length of zero?

I'm not sure if our library supports it...

Reply:
Looks like payload length of zero is not supported. I'm getting error 22 (EINVAL)